content/news/2023-10-21-bevy-0.12/index.md

```diff
@@ -19,6 +19,319 @@ Since our last release a few months ago we've added a _ton_ of new features, bug
+
+Instance-rate vertex buffers are one way, but they are very constrained to having a specific order. They are/may be suitable for per-instance data like mesh entity transforms, but they can't be used for material data. The other main options are uniform buffers, storage buffers, and data textures.
+
+WebGL2 does not support storage buffers, only uniform buffers. Uniform buffers have a minimum guaranteed size per binding of 16kB on WebGL2. Storage buffers, where available, have a minimum guaranteed size of 128MB.
+
+Data textures are far more awkward for structured data, and without support for linear data layouts on some platforms, they will perform worse.
+
+We want to support uniform buffers on WebGL2 or where storage buffers are not supported, and use storage buffers everywhere else.
+
+#### BatchedUniformBuffer
+
+<div class="release-feature-authors">authors: Rob Swain (@superdump), @JMS55, @teoxoy, @robtfm, @konsolas</div>
+
+We have to assume that on WebGL2, we may only be able to access 16kB of data at a time. Taking an example, `MeshUniform` requires 144 bytes per instance, which means we can have a batch of 113 instances per 16kB binding. If we want to draw more than 113 entities in total, we need a way of managing a uniform buffer of data that can be bound at a dynamic offset per batch of instances. This is what `BatchedUniformBuffer` is designed to solve.
+
+`BatchedUniformBuffer` looks like this:
+
+![BatchedUniformBuffer](BatchedUniformBuffer.svg)
+<div style="font-size: 1.0rem" class="release-feature-authors">Red arrows are 'rebinds' to update the dynamic offset, blue boxes are instance data, orange boxes are padding for dynamic offset alignment.</div>
+
+Notice how the instance data can be packed much more tightly, fitting the same amount of used data in less space. Also, we only need to update the dynamic offset of the binding for each batch.
+
+#### GpuArrayBuffer
+
+<div class="release-feature-authors">authors: Rob Swain (@superdump), @JMS55, @IceSentry, @mockersf</div>
+
+If users have to care about supporting both batched uniform and storage buffers to store arrays of data for use in shaders, many may choose not to because their priority is not WebGL2. We want to make it simple and easy to support all users.
+
+`GpuArrayBuffer` was designed and implemented as an abstraction over `BatchedUniformBuffer` and using a `StorageBuffer` to store an array of `T`.
+
+The data in a `StorageBuffer` looks like this:
+
+![StorageBuffer](StorageBuffer.svg)
+<div style="font-size: 1.0rem" class="release-feature-authors">Red arrows are 'rebinds', blue boxes are instance data.</div>
+
+All the instance data can be placed directly one after the other, and we only have to bind once. There is no need for any dynamic offset binding, so there is no need for any padding for alignment.
+
+```rust
```

Comment:
This code example would be good as documentation / Bevy Book content, but I think its just slightly too technical / verbose for the average reader. It cross cuts into a lot of different features (shaders, shader defs, ShaderType/encase, GpuArrayBuffer, pipeline layouts). I think the percentage of readers that will understand whats happening here is significantly less than %1. I think we can get away with cutting it.

Reply:
I would cut this. Too long and crunchy I think.